src/main/java/com/example/service/NasaPictureService.java

```diff
@@ -0,0 +1,70 @@
+package com.example.service;
+
+import com.example.dto.NasaResponse;
+import com.example.dto.Photo;
+import com.example.exception.NasaServiceException;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Optional;
+
+import static java.util.Comparator.comparing;
+
+@Service
+@RequiredArgsConstructor
+public class NasaPictureService {
+
+    private final RestTemplate restTemplate;
+    @Value("${nasa.api.url}")
+    private String nasaBaseUrl;
+    @Value("${nasa.api.key}")
+    private String nasaApiKey;
+
+    @Cacheable("largestPicture")
+    public Optional<Photo> findLargestPicture(Integer sol, Optional<String> camera) {
+        var roverPhotos = getCuriosityRoverPhotos(sol, camera);
+        return roverPhotos
+                .parallelStream()
+                .map(this::enrichPhotoWithSize)
+                .max(comparing(Photo::getSize));
+    }
+
+    private Photo enrichPhotoWithSize(Photo photo) {
+        var headers = restTemplate.headForHeaders(photo.getPhotoUrl());
+        URI location = headers.getLocation();
+
+        if (location != null) {
+            headers = restTemplate.headForHeaders(location);
+            photo.setSize(headers.getContentLength());
+        }
+
+        return photo;
+    }
+
+    private List<Photo> getCuriosityRoverPhotos(Integer sol, Optional<String> camera) {
+        try {
+            String nasaUrl = buildNasaUrl(sol, camera);
+            var response = restTemplate.getForEntity(nasaUrl, NasaResponse.class);
+
+            return response.getBody().photos();
```

Comment:
Possible NPE here, body can be null. I would do something like this 
`Optional.ofNullable(response.getBody())
                    .map(NasaResponse::photos).orElse(List.of());` . 
What do you think? 🤔

Reply:
In that case also would be good to add filter for not null values in the `findLargestPicture `